datafusion/optimizer/src/common_subexpr_eliminate.rs

```diff
@@ -242,12 +245,16 @@ impl OptimizerRule for CommonSubexprEliminate {
             }
         };
 
-        // add an additional projection if the output schema changed.
-        if optimized_plan.schema() != &original_schema {
-            optimized_plan = build_recover_project_plan(&original_schema, optimized_plan);
+        match optimized_plan {
+            Some(optimized_plan) if optimized_plan.schema() != &original_schema => {
+                // add an additional projection if the output schema changed.
+                Ok(Some(build_recover_project_plan(
+                    &original_schema,
+                    optimized_plan,
+                )))
+            }
+            plan => Ok(plan),
```

Comment:
You might be able to do this something like

```suggestion
        optimized_plan.map(|optimized_plan| {
            if optimized_plan.schema() != &original_schema => {
                // add an additional projection if the output schema changed.
                build_recover_project_plan(
                    &original_schema,
                    optimized_plan,
                )
            });
```

Not sure if that is all that much better

Reply:
Hmm, what is the rationale of this suggestion 🤔? It can't pass the type checker, I'm afraid.

Reply:
The rationale was to make it slightly less verbose and more "functional" by using `map` rather than `match`. i don't think it is required